used in both reg and classi,constructs best decis boundary(hyperplanea) in multi d space to separate different classes.

linear ,polynomial ,radial basis function kernel


In [200]:
import numpy as np
import pandas as pd

# Visualization (optional)
import matplotlib.pyplot as plt
import seaborn as sns

# Preprocessing
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler, LabelEncoder

# Models
from sklearn.linear_model import LogisticRegression
from sklearn.neighbors import KNeighborsClassifier
from sklearn.svm import SVC

# Metrics
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score


In [201]:
df = pd.read_csv("Social_Network_Ads.csv")
df.head()


User ID  Gender  Age  EstimatedSalary  Purchased
0  15624510    Male   19            19000          0
1  15810944    Male   35            20000          0
2  15668575  Female   26            43000          0
3  15603246  Female   27            57000          0
4  15804002    Male   19            76000          0

In [202]:
df.drop(columns=["User ID"], inplace=True)


In [203]:
df.duplicated().sum()


np.int64(20)

In [204]:
df.isnull().sum()


Gender             0
Age                0
EstimatedSalary    0
Purchased          0
dtype: int64

In [205]:
df.drop(columns=["User ID"], errors="ignore", inplace=True)


In [206]:
df.drop_duplicates(inplace=True)


In [207]:
le = LabelEncoder()
df["Gender"] = le.fit_transform(df["Gender"])


In [208]:
X = df.drop("Purchased", axis=1)
y = df["Purchased"]


In [209]:
X_train, X_test, y_train, y_test = train_test_split(
    X, y,
    test_size=0.30,
    random_state=42,
    stratify=y
)


In [210]:
scaler = StandardScaler()

X_train = scaler.fit_transform(X_train)
X_test = scaler.transform(X_test)


In [211]:
models = {
    "Logistic Regression": LogisticRegression(),
    "KNN": KNeighborsClassifier(n_neighbors=5),
    "SVM Linear": SVC(kernel="linear"),
    "SVM Polynomial": SVC(kernel="poly", degree=3),
    "SVM RBF": SVC(kernel="rbf")
}


In [212]:
results = []

for name, model in models.items():
    model.fit(X_train, y_train)
    y_pred = model.predict(X_test)

    results.append({
        "Model": name,
        "Test Accuracy": accuracy_score(y_test, y_pred),
        "Test Precision": precision_score(y_test, y_pred),
        "Test Recall": recall_score(y_test, y_pred),
        "Test F1-Score": f1_score(y_test, y_pred)
    })


In [213]:
results_df = pd.DataFrame(results)
results_df


Model  Test Accuracy  Test Precision  Test Recall  \
0  Logistic Regression       0.807018        0.738095     0.738095   
1                  KNN       0.885965        0.795918     0.928571   
2           SVM Linear       0.798246        0.743590     0.690476   
3       SVM Polynomial       0.824561        0.805556     0.690476   
4              SVM RBF       0.885965        0.795918     0.928571   

   Test F1-Score  
0       0.738095  
1       0.857143  
2       0.716049  
3       0.743590  
4       0.857143

best model is knn as recall score 0.93 with much simple model
